### PR TITLE
Iteration one

### PR DESCRIPTION
**Pull request by RenarsS**
Tasks that are done:

1. Algorithm of the game implemented.
2. Console app.
3. UI displays messages requiring user input for height and width of the board (field).
4. Iteration changes every second.

### CODE REVIEW

File: GameOfLife/GameOfLife/Board.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+namespace GameOfLife
+{
+    public class Board
+    {
+        /// <summary>
+        /// Two-dimensional array in which initial values are stored
+        /// and new layout created.
+        /// </summary>
+        private bool[,] initialBoard;
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs:11]: why is the summary only for this?

File: GameOfLife/GameOfLife/Board.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs:12]: no need for empty lines when there is no logical reason for it to be there, as all these variables are private

File: GameOfLife/GameOfLife/Board.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        private readonly int _height;
+        private readonly int _width;
+
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs:17]: leave only one line empty between code lines.

File: GameOfLife/GameOfLife/Board.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    public class Board
+    {
+        /// <summary>
+        /// Two-dimensional array in which initial values are stored
+        /// and new layout created.
+        /// </summary>
+        private bool[,] initialBoard;
+
+        private readonly int _height;
+        private readonly int _width;
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs]: Is there a need for height and width if you can get those values from the array itself?

File: GameOfLife/GameOfLife/Board.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            bool[,] newBoard = new bool[_height, _width];
+
+            for(int i = 0; i < initialBoard.GetLength(0); i++)
+            {
+                for (int j = 0; j < initialBoard.GetLength(1); j++)
+                {
+                    int aliveCellCount = Count(i, j);
+
+                    bool newCell = Determine(aliveCellCount, initialBoard[i, j]);
+                    newBoard[i,j] = newCell;
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs]: both lines can be written into one `newBoard[i,j] = Determine(aliveCellCount, initialBoard[i, j]);`

File: GameOfLife/GameOfLife/Board.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        /// Displays board on console.
+        /// Live cells are indicated by white square and un-alived cells by nothing.
+        /// </summary>
+        public void DisplayBoard()
+        {
+            for (int i = 0; i < initialBoard.GetLength(0); i++)
+            {
+                for (var j = 0; j < initialBoard.GetLength(1); j++)
+                {
+                    Console.Write(initialBoard[i, j] ? (char)009632 + " " : "  ");
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs]: why use `(char)009632` and no magic strings, `" "` can be defined as a const the same with the char used for alive cells.

File: GameOfLife/GameOfLife/Board.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Iterates through all of the neighbours of a given cell indices.
+        /// </summary>
+        /// <param name="x"></param>
+        /// <param name="y"></param>
+        /// <returns>Integer which represents amount of live cells</returns>
+        public int Count( int x, int y)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs]: Count what? Please more explanatory names for methods

File: GameOfLife/GameOfLife/Board.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            return aliveNeighbourCells;
+        }
+
+        /// <summary>
+        /// Determines state of the cell based on parameters and rules.
+        /// </summary>
+        /// <param name="neighbourCount">Amount of live neighbour cells.</param>
+        /// <param name="initialState">Indicates state of the cell prior to rule application.</param>
+        /// <returns>State of the cell with the same.</returns>
+        public bool Determine(int neighbourCount, bool initialState)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Board.cs]: Determine what?

File: GameOfLife/GameOfLife/Program.cs [hunk 9]
```diff
@@ -1,2 +1,17 @@
...
+string height = Console.ReadLine();
+
+Console.WriteLine("Enter width, please:");
+string width = Console.ReadLine();
+
+Console.Clear();
+
+var board = new Board(int.Parse(height), int.Parse(width));
+
+board.Play();
```
Comment (taurins) [inline GameOfLife/GameOfLife/Program.cs]: In general no logic in Program.cs file, only the call to the logic, preferable couple of lines

File: GameOfLife/GameOfLife/Board.cs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+            _height = height;
+            _width = width;
+
+            Random rnd = new Random();
+
+            for (int i = 0; i < height; i++)
+            {
+                for(int j = 0; j < width; j++)
+                {
+                    initialBoard[i, j] = rnd.Next(0, 2) == 1 ? true : false;
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Board.cs]: Can be simplified:
initialBoard[i, j] = rnd.Next(0, 2) == 1;

File: GameOfLife/GameOfLife/Board.cs (hunk 5 above)
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Board.cs]: Space after comma

File: GameOfLife/GameOfLife/Board.cs [hunk 11]
```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GameOfLife
+{
+    public class Board
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs:6]: Public class, please add comment.

File: GameOfLife/GameOfLife/Board.cs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+        private readonly int _height;
+        private readonly int _width;
+
+
+        /// <summary>
+        /// Creates two-dimensional array with user-desired sizes
+        /// and populates it with randomly generated boolean values.
+        /// </summary>
+        /// <param name="height"></param>
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs]: Missing param comment.

File: GameOfLife/GameOfLife/Board.cs [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        /// <param name="width"></param>
+        public Board(int height, int width)
+        {
+            initialBoard = new bool[height, width];
+
+            _height = height;
+            _width = width;
+
+            Random rnd = new Random();
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs]: What is rnd ?

File: GameOfLife/GameOfLife/Board.cs [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+        /// <summary>
+        /// Iterates whole board of cells one time. 
+        /// Results in newer generation of live cells based on rules.
+        /// </summary>
+        public void Iterate()
+        {
+            bool[,] newBoard = new bool[_height, _width];
+
+            for(int i = 0; i < initialBoard.GetLength(0); i++)
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs:80]: Why not use _height and _width ?

File: GameOfLife/GameOfLife/Board.cs [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        /// </summary>
+        public void DisplayBoard()
+        {
+            for (int i = 0; i < initialBoard.GetLength(0); i++)
+            {
+                for (var j = 0; j < initialBoard.GetLength(1); j++)
+                {
+                    Console.Write(initialBoard[i, j] ? (char)009632 + " " : "  ");
+                }
+                Console.WriteLine();
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs:106]: Please add new line after code block.

File: GameOfLife/GameOfLife/Board.cs [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Iterates through all of the neighbours of a given cell indices.
+        /// </summary>
+        /// <param name="x"></param>
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs]: Missing param comment.

Comment (ArvisZalitis) [reply GameOfLife/GameOfLife/Board.cs]: And what is x ?

File: GameOfLife/GameOfLife/Program.cs [hunk 17]
```diff
@@ -1,2 +1,17 @@
...
+
+Console.WriteLine("Enter height, please: ");
+string height = Console.ReadLine();
+
+Console.WriteLine("Enter width, please:");
+string width = Console.ReadLine();
+
+Console.Clear();
+
+var board = new Board(int.Parse(height), int.Parse(width));
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Program.cs]: Please validate user input, for example if user enter @ or -10

**Review by ArvisZalitis — CHANGES_REQUESTED** [review]

File: GameOfLife/GameOfLife/Board.cs (hunk 1 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs:11]: Summary added everywhere.

File: GameOfLife/GameOfLife/Board.cs (hunk 11 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs:6]: Comment added with short description.

File: GameOfLife/GameOfLife/Board.cs (hunk 13 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Descriptive comment added.

File: GameOfLife/GameOfLife/Board.cs (hunk 15 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs:106]: New lines added.

File: GameOfLife/GameOfLife/Program.cs (hunk 17 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Program.cs]: Input value validation enabled. Validation class created with appropriate method.

File: GameOfLife/GameOfLife/Board.cs (hunk 13 above)
Comment (ArvisZalitis) [reply GameOfLife/GameOfLife/Board.cs]: hmm I would say, update rnd variable to more meaningful name and remove comment.

Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Variable name changed.

File: GameOfLife/GameOfLife/Board.cs (hunk 16 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: x indicated the row and y indicated the column in which cell could be found but as it was not very understandable I changed them x to row and y to column. Param comment added.

File: GameOfLife/GameOfLife/Board.cs (hunk 5 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Space added.

File: GameOfLife/GameOfLife/Board.cs (hunk 10 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Code edited accordingly. Redundant part "? true : false" has been removed.

File: GameOfLife/GameOfLife/Board.cs (hunk 3 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs:17]: How to resolve this issue? I don't see what is incorrect. There is only one line after the property and the summary of another property.

Comment (taurins) [reply GameOfLife/GameOfLife/Board.cs:17]: line 19 and 20 is empty, there is no need to leave multiple empty lines, only one is ever needed.

File: GameOfLife/GameOfLife/Board.cs (hunk 12 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Comment added.

**Review by ArvisZalitis — APPROVED** [review]

File: GameOfLife/GameOfLife/Board.cs [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+            } while (true);
+
+        }
+
+        /// <summary>
+        /// Iterates whole board of cells one time. 
+        /// Results in newer generation of live cells based on rules.
+        /// </summary>
+        public void Iterate()
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs]: Why this method is public ?

File: GameOfLife/GameOfLife/Board.cs [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+        /// <summary>
+        /// Iterates through all of the neighbours of a given cell indices.
+        /// </summary>
+        /// <param name="row">Index of the row of the cell which neighbours are checked.</param>
+        /// <param name="column">Index of the cell in the row.</param>
+        /// <returns>Integer which represents amount of live cells.</returns>
+        public int CountLiveNeighbours( int currentRow, int currentColumn)
+        {
+            int aliveNeighbourCells = 0;
```
Comment (ArvisZalitis) [inline GameOfLife/GameOfLife/Board.cs]: Redundant assignment.

**Review by ArvisZalitis — APPROVED** [review]

**Review by taurins — APPROVED** [review]

File: GameOfLife/GameOfLife/Board.cs (hunk 18 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Changed to private.

File: GameOfLife/GameOfLife/Board.cs (hunk 19 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Assignment removed.

File: GameOfLife/GameOfLife/Board.cs (hunk 4 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Redundant properties removed.

File: GameOfLife/GameOfLife/Board.cs (hunk 5 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Code rewritten as suggested.

File: GameOfLife/GameOfLife/Board.cs (hunk 6 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Characters defined as constants.

File: GameOfLife/GameOfLife/Board.cs (hunk 7 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Changed to more concise name.

File: GameOfLife/GameOfLife/Board.cs (hunk 8 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Board.cs]: Changed to DetermineCellState.

File: GameOfLife/GameOfLife/Program.cs (hunk 9 above)
Comment (RenarsS) [reply GameOfLife/GameOfLife/Program.cs]: Created another class that encapsulates all of the logic.

**Pull request merged**